### PR TITLE
feat(functions): add firebase analytics sample

### PR DESCRIPTION
**Pull request by ace-n**
~**Do not merge** until this is tested against Firebase Analytics.~

**UPDATE:** @schandel and I confirmed that this works on Firebase Analytics.

(It's not possible to test this without having an Android application handy, so that's why I omitted the `DeployTest`.)

### CODE REVIEW

**Comment by ace-n** [comment]
@schandel can you confirm that this sample works? Thanks!

**Comment by schandel** [comment]
Tested sample, looks like the function is being triggered on a conversion event being logged, but the call to hello_analytics is failing. Will follow-up on email to investigate.

**Comment by ace-n** [comment]
The function name (as used in the sample) is `firebaseAnalytics`. I'm presuming that you used `hello_analytics` in your console-deployed function, though?

**Comment by schandel** [comment]
Hm... I actually didn't define a hello_analytics target function name. Maybe that's a default that the CF Analytics integration expects? I just granted you edit access to the test project to check it out.

**Comment by ace-n** [comment]
Update: got it working, thanks Sumit!

File: functions/firebase_analytics/composer.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "scripts": {
+        "start": [
+            "Composer\\Config::disableProcessTimeout",
+            "FUNCTION_SIGNATURE_TYPE=cloudevent FUNCTION_TARGET=firebaseAnalytics php -S localhost:${PORT:-8080} vendor/bin/router.php"
+        ]
+    },
+    "require-dev": {
+        "google/auth": "^1.14",
+        "google/cloud-logging": "^1.21"
```
Comment (bshaffer) [inline functions/firebase_analytics/composer.json:13]: `google/auth` is a dependency of `google/cloud-logging` and probably doesn't need to be defined individually

File: functions/firebase_analytics/test/IntegrationTest.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+                        'userDim' => [
+                            'geoInfo' => [
+                                'city' => 'San Francisco',
+                                'country' => 'US'
+                            ],
+                            'deviceInfo' => [
+                                'deviceModel' => 'Google Pixel XL'
+                            ]
+                        ],
+                        'updateType' => '🤔🤔🤔🤔',
```
Comment (bshaffer) [inline functions/firebase_analytics/test/IntegrationTest.php:65]: lol

File: functions/firebase_analytics/index.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START functions_firebase_analytics]
+
+use Google\CloudFunctions\CloudEvent;
+
+function firebaseAnalytics(CloudEvent $cloudevent)
```
Comment (bshaffer) [inline functions/firebase_analytics/index.php:22]: Add `void` return type for clarity

```suggestion
function firebaseAnalytics(CloudEvent $cloudevent): void
```

File: functions/firebase_analytics/index.php [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    $log = fopen(getenv('LOGGER_OUTPUT') ?: 'php://stderr', 'wb');
+    
+    $data = $cloudevent->getData();
+
+    fwrite($log, 'Function triggered by the following event:' . $data['resource'] . PHP_EOL);
+
+    $analyticsEvent = $data['eventDim'][0];
+    $unixTime = $analyticsEvent['timestampMicros'] / 1000;
+    ;
```
Comment (bshaffer) [inline functions/firebase_analytics/index.php:32]:
```suggestion
```

File: functions/firebase_analytics/index.php [hunk 5]
```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * Copyright 2020 Google LLC.
```
Comment (bshaffer) [inline functions/firebase_analytics/index.php:3]:
```suggestion
 * Copyright 2021 Google LLC.
```

File: functions/firebase_analytics/phpunit.xml.dist [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  Copyright 2020 Google LLC
```
Comment (bshaffer) [inline functions/firebase_analytics/phpunit.xml.dist:3]:
```suggestion
  Copyright 2021 Google LLC
```

File: functions/firebase_analytics/test/IntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
+<?php
+/**
+ * Copyright 2020 Google LLC.
```
Comment (bshaffer) [inline functions/firebase_analytics/test/IntegrationTest.php:3]:
```suggestion
 * Copyright 2021 Google LLC.
```

**Review by bshaffer — COMMENTED** [review]
This looks great but... no `Deploy` test?

File: functions/firebase_analytics/composer.json (hunk 1 above)
Comment (bshaffer) [reply functions/firebase_analytics/composer.json:13]: On second glance, we aren't using `google/cloud-logging` either, so this whole block can be removed

File: functions/firebase_analytics/test/IntegrationTest.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        foreach ($data['eventDim'] as $property => $value) {
+            if (is_string($value)) {
+                $this->assertStringContainsString($value, $actual);
+            }
+        }
+        foreach ($data['userDim'] as $property => $value) {
+            if (is_string($value)) {
+                $this->assertStringContainsString($value, $actual);
+            }
```
Comment (bshaffer) [inline functions/firebase_analytics/test/IntegrationTest.php:106]: We should be able to just cast `$value` to string in both cases:

```php
        foreach ($data['eventDim'] as $property => $value) {
            $this->assertStringContainsString((string) $value, $actual);
        }
        foreach ($data['userDim'] as $property => $value) {
            $this->assertStringContainsString((string) $value, $actual);
        }
```

File: functions/firebase_analytics/test/IntegrationTest.php (hunk 2 above)
Comment (ace-n) [reply functions/firebase_analytics/test/IntegrationTest.php:65]: @schandel happy to put an actual value here if you'd like 😝

File: functions/firebase_analytics/index.php (hunk 3 above)
Comment (ace-n) [reply functions/firebase_analytics/index.php:22]: Done.

File: functions/firebase_analytics/index.php (hunk 4 above)
Comment (ace-n) [reply functions/firebase_analytics/index.php:32]: Good catch, done.

File: functions/firebase_analytics/phpunit.xml.dist (hunk 6 above)
Comment (ace-n) [reply functions/firebase_analytics/phpunit.xml.dist:3]: Done.

File: functions/firebase_analytics/test/IntegrationTest.php (hunk 7 above)
Comment (ace-n) [reply functions/firebase_analytics/test/IntegrationTest.php:3]: Done.

File: functions/firebase_analytics/test/IntegrationTest.php (hunk 8 above)
Comment (ace-n) [reply functions/firebase_analytics/test/IntegrationTest.php:106]: Some of these are arrays, which can't be casted to strings. (That returns an `Array to string conversion` error message.)

**Comment by ace-n** [comment]
RE the lack of `DeploymentTest`s - I was debating whether or not to document this in the repo, but there is no way to trigger these functions without using a mobile app, as the Firebase "web" SDK events **do not** trigger these functions!

(@schandel is aware of this; IDK if the Firebase team will attempt to change it or not. cc @samtstern as FYI.)

Emulators work for this too, but do we _really_ want to add a mobile emulator to our toolchain (especially for a single sample)? 🙃

File: functions/firebase_analytics/index.php (hunk 5 above)
Comment (ace-n) [reply functions/firebase_analytics/index.php]: Done.

File: functions/firebase_analytics/composer.json (hunk 1 above)
Comment (ace-n) [reply functions/firebase_analytics/composer.json]: Good catch - done.

File: functions/firebase_analytics/composer.json [hunk 9]
```diff
@@ -0,0 +1,11 @@
+{
+    "require": {
+        "google/cloud-functions-framework": "^0.7.1"
```
Comment (bshaffer) [inline functions/firebase_analytics/composer.json:3]: nit: update to `^0.7.2`
```suggestion
        "google/cloud-functions-framework": "^0.7.2"
```

**Review by bshaffer — APPROVED** [review]

File: functions/firebase_analytics/index.php [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    
+    $data = $cloudevent->getData();
+
+    fwrite($log, 'Function triggered by the following event:' . $data['resource'] . PHP_EOL);
+
+    $analyticsEvent = $data['eventDim'][0];
+    $unixTime = $analyticsEvent['timestampMicros'] / 1000;
+
+    fwrite($log, 'Name: ' . $analyticsEvent['name'] . PHP_EOL);
+    fwrite($log, 'Timestamp: ' . gmdate("Y-m-d\TH:i:s\Z", $unixTime) . PHP_EOL);
```
Comment (grayside) [inline functions/firebase_analytics/index.php:34]: Is there potential for confusion between this timestamp and the CloudEvent timestamp?

File: functions/firebase_analytics/index.php [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+    $userObj = $data['userDim'];
+    fwrite($log, sprintf(
+        'Location: %s, %s' . PHP_EOL,
+        $userObj['geoInfo']['city'],
+        $userObj['geoInfo']['country']
+    ));
+
+    fwrite($log, 'Device Model: %s' . $userObj['deviceInfo']['deviceModel'] . PHP_EOL);
```
Comment (grayside) [inline functions/firebase_analytics/index.php:43]: I appreciate we're not logging the location along with device model, as that would be a potential privacy risk if used in practice.

**Review by grayside — APPROVED** [review]

**Comment by ace-n** [comment]
@schandel @samtstern I added y'all as code owners for this sample, since we (Cloud DPE) don't have the ability to test it ourselves (due to the "must have an app" requirement).

Does that sound OK to you?

File: functions/firebase_analytics/index.php (hunk 11 above)
Comment (ace-n) [reply functions/firebase_analytics/index.php:43]: You're saying that location and device model might be used for [fingerprinting](https://en.wikipedia.org/wiki/Device_fingerprint)?

(Right now, I _believe_ the logs can be correlated within Cloud Logging - so I'm not sure logging them separately absolves this risk.)

File: functions/firebase_analytics/index.php (hunk 10 above)
Comment (ace-n) [reply functions/firebase_analytics/index.php:34]: Maybe. @grant is there somewhere we can link to that explains the default CloudEvent parameters?

**Pull request merged**